odohlistener.go

```diff
@@ -0,0 +1,250 @@
+	addr        string
+	proxyClient *http.Client // Client for proxy forwarding if acting as ODoH proxy
+	doh         *DoHListener
+
+	r           Resolver // Forwarding DNS queries if acting as ODoH Target
+	opt         ODoHListenerOptions
+	odohKeyPair odoh.ObliviousDoHKeyPair
+}
+
+type ODoHListenerOptions struct {
+	ListenOptions
+
+	OdohMode  string
+	AllowDoH  bool
+	KeySeed   string
+	TLSConfig *tls.Config
+}
+
+var _ Listener = &ODoHListener{}
+
+// NewODoHListener returns an instance of an oblivious DNS-over-HTTPS listener.
+func NewODoHListener(id, addr string, opt ODoHListenerOptions, resolver Resolver) (*ODoHListener, error) {
+	keyPair, err := getKeyPair(opt)
+	if err != nil {
+		Log.Error("Failed to generate HPKE key pair", "error", err)
+		return nil, err
+	}
+
+	dohOpt := DoHListenerOptions{
+		TLSConfig: opt.TLSConfig,
+		isChild:   true,
+	}
+	dohListen, err := NewDoHListener(id, addr, dohOpt, resolver)
+	if err != nil {
+		Log.Error("Failed to spawn DoH listener", "error", err)
+		return nil, err
+	}
+
+	l := &ODoHListener{
+		id:          id,
+		addr:        addr,
+		r:           resolver,
+		opt:         opt,
+		proxyClient: &http.Client{},
+		odohKeyPair: keyPair,
+		doh:         dohListen,
+	}
+
+	switch opt.OdohMode {
+	case "proxy":
+		http.HandleFunc(ODOH_PROXY_PATH, l.ODoHproxyHandler)
+	case "target":
+		http.HandleFunc(ODOH_QUERY_PATH, l.ODoHqueryHandler)
+		http.HandleFunc(ODOH_CONFIG_PATH, l.configHandler)
+	default:
+		http.HandleFunc(ODOH_PROXY_PATH, l.ODoHproxyHandler)
+		http.HandleFunc(ODOH_QUERY_PATH, l.ODoHqueryHandler)
+		http.HandleFunc(ODOH_CONFIG_PATH, l.configHandler)
+	}
+	return l, nil
+}
+
+func (s *ODoHListener) Start() error {
+	return s.doh.Start()
+}
+
+func (s *ODoHListener) Stop() error {
+	return s.doh.Stop()
+}
+
+func (s *ODoHListener) ODoHproxyHandler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		http.Error(w, "only POST allowed", http.StatusMethodNotAllowed)
+		return
+	}
+	contentType := r.Header.Get("Content-Type")
+	host := r.URL.Query().Get("targethost")
+	if host == "" {
+		http.Error(w, "no targethost specified", http.StatusBadRequest)
+		return
+	}
+	path := r.URL.Query().Get("targetpath")
+	if path == "" {
+		http.Error(w, "no targetpath specified", http.StatusBadRequest)
+		return
+	}
+
+	defer r.Body.Close()
+	b, err := io.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	Log.Debug("forwarding query to ODoH target",
+		slog.String("client", r.RemoteAddr),
+		slog.String("target", host),
+	)
+	response, err := forwardProxyRequest(s.proxyClient, host, path, b, contentType)
+	if err != nil {
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+	if response.StatusCode != 200 {
+		http.Error(w, http.StatusText(response.StatusCode), response.StatusCode)
+		return
+	}
+
+	defer response.Body.Close()
+	rb, err := io.ReadAll(response.Body)
```

Comment:
Same issue as above with this getting read fully into memory. The fix trivial in this case though. Just stream the response from the target through to the client like so:
```go
w.Header().Set("Content-Type", contentType) // Probably want to use the Content-Type received from the target

io.Copy(w, response.Body)
```